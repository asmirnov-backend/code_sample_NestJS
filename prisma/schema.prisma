generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["filterJson"]
}

datasource db {
  provider = "mysql"
  url      = env("DATABASE_URL")
}

model Advertiser {
  id        BigInt     @id @default(autoincrement())
  originId  String     @unique @db.VarChar(255)
  name      String     @db.VarChar(255)
  campaigns Campaign[]
}

model Campaign {
  id              BigInt        @id @default(autoincrement())
  originId        Int           @unique
  name            String        @db.VarChar(255)
  friendlyName    String        @db.VarChar(255)
  contractStart   String?       @db.VarChar(255)
  contractEnd     String?       @db.VarChar(255)
  impressionsGoal Int?
  timezone        String?       @db.VarChar(255)
  spendRate       Decimal       @default(0)
  spendType       SpendType     @default(Hide)
  type            AnalyticsType
  advertiserId    BigInt?
  advertiser      Advertiser?   @relation(fields: [advertiserId], references: [id], onDelete: NoAction, onUpdate: NoAction, map: "FK_campaign_advertiser")

  @@index([advertiserId], map: "FK_campaign_advertiser")
}

enum SpendType {
  Hide
  CPM
}

enum AnalyticsType {
  AMAZONOTT
  AUDIO
  BROADCAST
  BROADSTREET
  CALLTRACKING
  DISPLAY
  FBINSIGHTS
  GA
  GAMDISPLAY
  GAMVIDEO
  GOOGLESEARCH
  GOOGLEVIDEO
  OTT
  PREROLL
  SEM
  SIMPGEOFENCE
  SITEIMPACT
  SOCIAL
  VIDEO
}

model broadstreet_performance_campaign {
  campaign_id  BigInt
  date         DateTime @db.Date()
  network_id   BigInt?
  views        BigInt?
  hovers       BigInt?
  clicks       BigInt?
  ctr          Float?   @db.Double
  total_views  BigInt?
  total_hovers BigInt?
  total_clicks BigInt?
  total_ctr    Float?   @db.Double

  @@id([campaign_id, date])
}

model broadstreet_performance_creatives {
  date                DateTime  @db.Date()
  campaign_id         BigInt
  creative_name       String    @db.VarChar(255)
  creative_id         BigInt
  creative_start_date DateTime? @db.Date()
  views               BigInt?
  hovers              BigInt?
  clicks              BigInt?
  ctr                 Float?    @db.Double
  total_views         BigInt?
  total_hovers        BigInt?
  total_clicks        BigInt?
  total_ctr           Float?    @db.Double

  @@id([campaign_id, date, creative_id])
}

model broadstreet_static_campaign_info {
  network_id      BigInt
  network_name    String?   @db.VarChar(255)
  advertiser_id   BigInt
  advertiser_name String?   @db.VarChar(255)
  campaign_name   String?   @db.VarChar(255)
  campaign_id     BigInt
  campaign_start  DateTime? @db.Date()
  campaign_end    DateTime? @db.Date()

  @@id([network_id, advertiser_id, campaign_id])
}
