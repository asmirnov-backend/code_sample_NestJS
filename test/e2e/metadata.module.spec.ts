import { HttpServer, INestApplication } from '@nestjs/common';
import { TestingModule, Test } from '@nestjs/testing';
import { GraphQLFormattedError } from 'graphql';
import * as request from 'supertest';

import { AppModule } from '@api/app.module';
import { Advertiser } from '@api/graphql.schema';

jest.setTimeout(10_000);
const GRAPHQL_URL = '/graphql';

describe('AdvertiserResolver', () => {
  let app: TestingModule;
  let nestApp: INestApplication;
  let server: HttpServer;

  beforeAll(async () => {
    app = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();

    nestApp = app.createNestApplication();
    await nestApp.init();
    server = nestApp.getHttpServer();
  });

  afterAll(async () => await nestApp.close());

  it('AdvertiserResolver', async () => {
    const result = await request(server).post(GRAPHQL_URL).send({
      query: '{ advertiser(originId: "1") { name id campaigns { name } }}',
    });

    const data: {
      data: Advertiser;
      error: { errors: GraphQLFormattedError[] };
    } = result.body;

    expect(data).not.toHaveProperty('error');
    expect(data.data).toMatchSnapshot();
  });
});
