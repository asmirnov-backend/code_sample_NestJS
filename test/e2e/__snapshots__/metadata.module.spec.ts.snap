// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdvertiserResolver AdvertiserResolver 1`] = `
Object {
  "advertiser": Object {
    "campaigns": Array [
      Object {
        "name": "name 1",
      },
    ],
    "id": 1,
    "name": "advertiser 1",
  },
}
`;
